### PR TITLE
BUG: fix IntervalIndex for pivot table raise type error 

### LINKED ISSUE #25814: Cannot use categorical IntervalIndex as index when creating pivot_table
#### Code Sample, a copy-pastable example if possible

```python
df = pd.DataFrame({"A":[1, 1, 2, 2], "B":[1,2,3,4]})
df['C'] = pd.cut(df['A'],[0,2])
df.pivot_table(index='C', values='B', aggfunc=np.sum)

```
#### Problem description
Cannot use IntervalIndex as index when creating  pivot_table. It works fine on 0.23.4.

C:\ProgramData\Anaconda3\lib\site-packages\pandas\core\indexes\interval.py in _get_prev_label(label)
     75     else:
     76         raise TypeError('cannot determine next label for type {typ!r}'
---> 77                         .format(typ=type(label)))
     78 
     79 

TypeError: cannot determine next label for type <class 'str'>

#### Output of ``pd.show_versions()``

<details>
INSTALLED VERSIONS
------------------
commit: None
python: 3.7.1.final.0
python-bits: 64
OS: Windows
OS-release: 10
machine: AMD64
processor: Intel64 Family 6 Model 142 Stepping 9, GenuineIntel
byteorder: little
LC_ALL: None
LANG: zh_CN.GBK@cjknarrow
LOCALE: None.None

pandas: 0.24.2
pytest: 4.0.2
pip: 18.1
setuptools: 40.6.3
Cython: 0.29.2
numpy: 1.15.4
scipy: 1.1.0
pyarrow: None
xarray: None
IPython: 7.2.0
sphinx: 1.8.2
patsy: 0.5.1
dateutil: 2.7.5
pytz: 2018.7
blosc: None
bottleneck: 1.2.1
tables: 3.4.4
numexpr: 2.6.8
feather: None
matplotlib: 3.0.2
openpyxl: 2.5.12
xlrd: 1.2.0
xlwt: 1.3.0
xlsxwriter: 1.1.2
lxml.etree: 4.2.5
bs4: 4.6.3
html5lib: 1.0.1
sqlalchemy: 1.2.15
pymysql: None
psycopg2: None
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None
gcsfs: Nonetput of ``pd.show_versions()`` here below this line]

</details>

### PR DESCRIPTION
**Pull request by peterpanmj**
Close (#25814). When use IntervalIndex for pivot table, raise
TypeError: cannot determine next label for type <class 'str'> 
- [x] closes #25814
- [x] tets added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/tests/reshape/test_pivot.py [hunk 1]
```diff
@@ -198,6 +198,17 @@ def test_pivot_with_non_observable_dropna(self, dropna):
 
         tm.assert_frame_equal(result, expected)
 
+    def test_pivot_with_interval_index(self, dropna):
+        df = pd.DataFrame(
```
Comment (jreback) [inline pandas/tests/reshape/test_pivot.py]: can you add the issue number as a comment

File: pandas/tests/reshape/test_pivot.py [hunk 2]
```diff
@@ -198,6 +198,17 @@ def test_pivot_with_non_observable_dropna(self, dropna):
...
+            {'A': pd.Categorical([pd.Interval(0, 1)] * 4),
+             'B': [1] * 4})
```
Comment (jreback) [inline pandas/tests/reshape/test_pivot.py]: does this fully replicate the OP issue?

**Review by jreback — CHANGES_REQUESTED** [review]
can you add a whatsnew note (reshaping) bug fixes

File: pandas/tests/reshape/test_pivot.py (hunk 2 above)
Comment (peterpanmj) [reply pandas/tests/reshape/test_pivot.py]: I just added a fixture to include more scenarios.

File: doc/source/whatsnew/v0.24.2.rst [hunk 3]
```diff
@@ -59,6 +59,7 @@ Bug Fixes
 
 - Bug in :meth:`~pandas.core.groupby.GroupBy.transform` where applying a function to a timezone aware column would return a timezone naive result (:issue:`24198`)
 - Bug in :func:`DataFrame.join` when joining on a timezone aware :class:`DatetimeIndex` (:issue:`23931`)
+- Bug in :func:`DataFrame.pivot_table` where use :class:`IntervalIndex` as pivot index would raise TypeError (:issue:`25814`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.24.2.rst]: can you use double-backticks on TypeError

**Review by jreback — CHANGES_REQUESTED** [review]
tiny tiny comment, ping on green.

**Review by jreback — APPROVED** [review]

File: pandas/tests/reshape/test_pivot.py [hunk 4]
```diff
@@ -19,6 +19,18 @@ def dropna(request):
...
 
+@pytest.fixture(
+    params=[
+        pd.Categorical([pd.Interval(0, 1, 'right')] * 4),
+        pd.Categorical([pd.Interval(0, 1, 'left')] * 4),
+        pd.Categorical([pd.Interval(low, high, 'right')
+                        for low, high in zip(range(0, 3), range(1, 4))]),
+        pd.Categorical([pd.Interval(low, high, 'left')
+                        for low, high in zip(range(0, 3), range(1, 4))])])
+def interval_values(request):
```
Comment (jschendel) [inline pandas/tests/reshape/test_pivot.py]: A bit cleaner implementation that tests all variants of closed:
```python
@pytest.fixture(params=[([0] * 4, [1] * 4), (range(0, 3), range(1, 4))])
def interval_values(request, closed):
    left, right = request.param
    return Categorical(pd.IntervalIndex.from_arrays(left, right, closed))
```

File: pandas/tests/reshape/test_pivot.py [hunk 5]
```diff
@@ -198,6 +210,18 @@ def test_pivot_with_non_observable_dropna(self, dropna):
 
         tm.assert_frame_equal(result, expected)
 
+    def test_pivot_with_interval_index(self, interval_values, dropna):
+        # GH 25814
+        df = pd.DataFrame(
+            {'A': interval_values,
+             'B': [1] * interval_values.size})
```
Comment (jschendel) [inline pandas/tests/reshape/test_pivot.py]: You don't need to match size here, can just do:
```python
df = DataFrame({'A': interval_values, 'B': 1})
```

File: pandas/tests/reshape/test_pivot.py [hunk 6]
```diff
@@ -198,6 +210,18 @@ def test_pivot_with_non_observable_dropna(self, dropna):
...
+        result = df.pivot_table(index='A', values='B', dropna=dropna)
+        expected = pd.DataFrame(
+            {'B': [1]},
+            index=pd.Index(interval_values.unique(),
```
Comment (jschendel) [inline pandas/tests/reshape/test_pivot.py]: Note that `DataFrame` and `Index` are both already imported, so you can omit the `pd.`

**Review by jschendel — COMMENTED** [review]
lgtm, just some simplifications to the implementation

File: doc/source/whatsnew/v0.24.2.rst [hunk 7]
```diff
@@ -59,7 +59,7 @@ Bug Fixes
 
 - Bug in :meth:`~pandas.core.groupby.GroupBy.transform` where applying a function to a timezone aware column would return a timezone naive result (:issue:`24198`)
 - Bug in :func:`DataFrame.join` when joining on a timezone aware :class:`DatetimeIndex` (:issue:`23931`)
-- Bug in :func:`DataFrame.pivot_table` where use :class:`IntervalIndex` as pivot index would raise `TypeError` (:issue:`25814`)
+- Bug in :func:`DataFrame.pivot_table` where use :class:`IntervalIndex` as pivot index would raise ``TypeError`` (:issue:`25814`)
```
Comment (TomAugspurger) [inline doc/source/whatsnew/v0.24.2.rst]: Change "where use" -> "with a"

File: doc/source/whatsnew/v0.24.2.rst (hunk 3 above)
Comment (peterpanmj) [reply doc/source/whatsnew/v0.24.2.rst]: Should I update the v0.25.0 whatsnew instead of v0.24.2 ?

File: doc/source/whatsnew/v0.24.2.rst (hunk 7 above)
Comment (peterpanmj) [reply doc/source/whatsnew/v0.24.2.rst]: Ok. I just realized this entry should be added to v0.25.0 whatsnew, right ?

Comment (jschendel) [reply doc/source/whatsnew/v0.24.2.rst]: yes, this should be in the 0.25.0 whatsnew

File: pandas/tests/reshape/test_pivot.py [hunk 8]
```diff
@@ -19,6 +19,17 @@ def dropna(request):
     return request.param
 
 
+@pytest.fixture(params=['right', 'left'])
+def closed(request):
```
Comment (jschendel) [inline pandas/tests/reshape/test_pivot.py]: There is a `closed` fixture in the top level `conftest.py` that's generally used with `IntervalIndex` that you can use instead.  It also includes `'both'` and `'neither'` as parameters.

Comment (peterpanmj) [reply pandas/tests/reshape/test_pivot.py]: I can't find it in
pandas/tests/indexing/conftest.py
or
pandas/tests/indexes/conftest.py
Can you give me a path to it ?

Comment (jschendel) [reply pandas/tests/reshape/test_pivot.py]: It's in `pandas/conftest.py`:
https://github.com/pandas-dev/pandas/blob/129aaea46db94d5f8296bd5daeff9dc124b1a99e/pandas/conftest.py#L337-L342

No need to import it or anything though; if you just delete the your current `closed` fixture `pytest` should automatically pick it up.

**Review by TomAugspurger — APPROVED** [review]
Merged master to fix the CI. Should be good to go.

**Comment by TomAugspurger** [comment]
Thanks @peterpanmj!

**Pull request merged**